docs/docs/debugging-incremental-builds.md

```diff
@@ -0,0 +1,98 @@
+
+## How does it work?
+
+To be able to understand why you might see more generated pages than expected it's helpful to know how Gatsby tracks "inputs" when generating HTML files. In particular Gatsby tracks:
+
+- which page template the page is using
+- results of page query
+- results of static queries used by a page template
+- frontend source code (shared and also browser `gatsby-browser.js` / SSR `gatsby-ssr.js` specifically)
+
+When those inputs change since the last build, the HTML files are marked to be regenerated. If they don't change, Gatsby can reuse HTML files generated in previous build.
+
+### Expected behaviors
+
+There are certain changes that you can make to your project that trigger more changes than anticipated or even a full rebuild. For the following cases this is expected behavior:
+
+- Changing a React component / page. This results in a changed webpack compilation where some assets generated by webpack will have changed hashes in filenames and thus Gatsby needs to regenerate pages to update `<link>` and `<script>` references.
+- Changing content queried by a static query. This results in one or more `.html` files being regenerated. In case of static query used by all page templates it would regenerate all pages (for example static query used by shared layout component)
+
+## Debugging
+
+To be able to see the changed pages inside console, run `gatsby build` with some additional flags:
+
+```shell
+gatsby build --verbose --log-pages
+```
+
+At the end of the build you will see the changed pages:
+
+```shell
+info Built pages:
+Updated page: /foo/bar/
+```
+
+If you also want to see what and how files actually did change between builds you can look at the `diff` between two builds. Inside your project directory run these commands:
+
+```shell
+gatsby clean
+gatsby build --no-uglify
+cp -r public public-first-build
+```
+
+If you are debugging why some changes result in more `.html` files being regenerated than expected, this is the time to make your changes to the files. If you are debugging why builds always regenerate all `.html` files, skip making any changes.
+
+Continue with these commands:
+
+```shell
+gatsby build --no-uglify --verbose --log-pages
+diff -u -r public-first-build public > build-diff.diff
+```
+
+Open the `build-diff.diff` file in the root of your site to see the difference between both builds.
+
+> **Note:** To exclude all `.html` files in the diff you can run this command at the end instead: `diff -u -r --exclude="*.html" public-first-build public > build-diff.diff`
+
+### Understanding the `diff`
+
+Generating the `diff` between two builds is only the first step, understanding and interpreting the `diff` is the second step. You can use tools like [diffchecker.com](https://www.diffchecker.com/) to see the differences more easily (or your code editor might have a feature/plugin for that).
+
+TODO
```

Comment:
@pieh the question here now is in what form we include https://github.com/gatsbyjs/gatsby/issues/30202#issuecomment-801839746 here :)